### ISSUE TITLE
[BEVY] NPC Creatures — Performance Audit

### ISSUE THREAD

**Issue by h0lybyte** [KBVE/kbve#8189]
## Overview

Performance audit of the `bevy_npc` and `bevy_kbve_net::npcdb` creature system covering memory leaks, unnecessary clones, hashing stability, broadcast amplification, and zero-copy opportunities.

Ref: #8152

---

### Box::leak() unbounded memory in NpcDb::insert()

**File:** `packages/rust/bevy/bevy_npc/src/registry.rs:67`

```rust
let name: &'static str = Box::leak(npc.name.clone().into_boxed_str());
```

- Leaks a heap-allocated String on every NPC insertion — never freed
- Double allocation: clones the String, then boxes it, then leaks
- Currently only 3 NPCs, but any dynamic loading or hot-reloading grows unbounded

**Fix:** Replace `display_names: HashMap<ProtoNpcId, &'static str>` with `HashMap<ProtoNpcId, Arc<str>>` or `HashMap<ProtoNpcId, String>`. Zero-copy read via `.as_str()`.

### Broadcast message cloning — O(N) clones per event

**File:** `apps/kbve/axum-kbve/src/gameserver/mod.rs` lines 767, 800, 879, 1033

```rust
for mut sender in &mut senders {
    sender.send::<GameChannel>(broadcast.clone());  // cloned per connected client
}
```

Affects: `ObjectRemoved`, `ObjectRespawned`, `CreatureCaptured`, `TimeSyncMessage`

- Each broadcast clones the message struct N times (once per connected client)
- `TimeSyncMessage` fires every 5 seconds — 100 clients = 100 clones/5s
- Capture/collect events are bursty but compound under load

**Fix options:**
1. Wrap broadcast messages in `Arc<T>` — single allocation, shared read
2. Use lightyear's broadcast API if available (send-once, replicate to all)
3. Offload serialization to **bevy_tasker** worker pool (see tooling section below)
4. For `TimeSyncMessage`: send on `TimeChannel` (unreliable) — already does, but still clones

---

### ProtoNpcId uses DefaultHasher — not stable across runs

**File:** `packages/rust/bevy/bevy_npc/src/registry.rs:17-23`

```rust
let mut h = std::collections::hash_map::DefaultHasher::new();
r.hash(&mut h);
Self(h.finish())
```

- `DefaultHasher` is SipHash with random seed — **not stable across process restarts**
- ProtoNpcId values will differ between server and client if they're in different processes
- Currently works because both sides call `ProtoNpcId::from_ref()` at runtime, but fragile for persistence (save files, database storage)

**Fix:** Use a deterministic hash like FNV-1a or xxHash:
```rust
pub fn from_ref(r: &str) -> Self {
    let mut h: u64 = 0xcbf29ce484222325; // FNV offset basis
    for b in r.as_bytes() {
        h ^= *b as u64;
        h = h.wrapping_mul(0x100000001b3); // FNV prime
    }
    Self(h)
}
```

### CapturedCreatures full iteration on client join

**File:** `apps/kbve/axum-kbve/src/gameserver/mod.rs` — `send_captured_state_to_new_clients()`

```rust
for &(npc_id, creature_index) in captured.iter() {
    sender.send::<GameChannel>(...);
}
```

- Iterates all captured creatures per new client join
- Scales as O(captured × joins) over session lifetime
- 1000 captured × 100 joins = 100k message sends

**Fix:** Pre-build a batch message or snapshot. Send a single `CapturedCreaturesBatch { entries: Vec<(kind, index)> }` instead of N individual messages.

---

### String cloning in NpcDb::insert()

**File:** `packages/rust/bevy/bevy_npc/src/registry.rs:69-71`

```rust
self.by_ref.insert(npc.r#ref.clone(), id);
if !npc.id.is_empty() {
    self.by_ulid.insert(npc.id.clone(), id);
}
```

- Two String clones per NPC, then the original `npc` is moved into `by_id`
- Could extract refs before the move or use `Arc<str>` for shared ownership

**Fix:** Extract strings before consuming the NPC:
```rust
let ref_key = npc.r#ref.clone();
let ulid_key = if npc.id.is_empty() { None } else { Some(npc.id.clone()) };
self.by_id.insert(id, npc); // moved
self.by_ref.insert(ref_key, id);
if let Some(ulid) = ulid_key { self.by_ulid.insert(ulid, id); }
```

Or better: use `Cow<'static, str>` / `Arc<str>` for NPC string fields to share without cloning.

### find_by_* methods allocate Vec on every call

**File:** `packages/rust/bevy/bevy_npc/src/registry.rs:104-133`

```rust
pub fn find_by_type_flags(&self, flags: i32) -> Vec<&npc::Npc> {
    self.by_id.values().filter(...).collect()
}
```

- Returns `Vec<&Npc>` — heap allocation on every query
- Called at startup only (currently), but bad API pattern

**Fix:** Return `impl Iterator<Item = &Npc>` instead:
```rust
pub fn find_by_type_flags(&self, flags: i32) -> impl Iterator<Item = &npc::Npc> {
    self.by_id.values().filter(move |npc| npc.type_flags & flags == flags)
}
```

### build_creature_registry() allocates strings with .into()

**File:** `packages/rust/bevy/bevy_kbve_net/src/npcdb.rs:140-284`

```rust
r#ref: "meadow-firefly".into(),   // &str → String allocation
name: "Meadow Firefly".into(),     // &str → String allocation
zone: Some("grassland".into()),    // &str → String allocation
```

- ~9 String allocations per creature (3 creatures = 27 allocations at startup)
- All from `&'static str` literals — could use `Cow::Borrowed` to avoid heap

**Fix:** Change proto Npc string fields to `Cow<'static, str>` or accept that startup cost is negligible. Low priority.

---

### Zero-copy creature pool queries

Current creature matching in `receive_creature_captured()` does a linear scan:
```rust
for (mut creature, pool_idx, mut vis) in &mut creatures {
    if creature.render_kind == render_kind && pool_idx.0 == msg.creature_index { ... }
}
```

- O(pool_size) per capture message — up to 102 entities (80 fireflies + 14 butterflies + 8 frogs)

**Fix options:**
1. Maintain a `HashMap<(RenderKind, u32), Entity>` index resource — O(1) lookup
2. Use Bevy's `Query` with a marker component per creature type to narrow the scan
3. Accept the linear scan (102 entities is trivially fast)

### Slot helpers — already optimal

`hash_f32()`, `slot_seed()`, `slot_anchor()`, `slot_active()` are pure functions with zero allocations. Bitwise wrapping arithmetic — no performance concerns. These run per-frame for chunk-based assignment and are well-optimized.

---

### DashMap — concurrent HashMap for shared mutable state

Standard Bevy `Res<T>`/`ResMut<T>` access is already synchronized by the ECS scheduler, so DashMap doesn't help for normal resource access. It becomes useful when **bevy_tasker** workers need concurrent read/write access to game state alongside the main thread.

**Where DashMap fits:**

| Use Case | Current | With DashMap |
|----------|---------|--------------|
| `CapturedCreatures` — worker validates capture while main thread reads | `ResMut<CapturedCreatures>` blocks | `DashMap<(ProtoNpcId, u32), ()>` — lock-free concurrent insert/contains |
| `NpcDb` — async loading from proto/JSON while systems query | Must block on `ResMut` | `DashMap<ProtoNpcId, Npc>` + `DashMap<String, ProtoNpcId>` — workers write, systems read |
| Creature pool index — O(1) entity lookup | Linear scan of Query | `DashMap<(RenderKind, u32), Entity>` — shared index, updated by spawn systems, read by capture handler |

**Where DashMap does NOT help:**
- Broadcast clone problem (lightyear's send API, not a data structure issue)
- Slot helpers (pure functions, no shared state)
- `CreatureRegistry` (read-only after startup — `Res<T>` is fine)

**Implementation pattern:**
```rust
use dashmap::DashMap;

#[derive(Resource, Default)]
struct CapturedCreatures {
    captured: DashMap<(ProtoNpcId, u32), ()>,
}

impl CapturedCreatures {
    fn insert(&self, npc_id: ProtoNpcId, idx: u32) { self.captured.insert((npc_id, idx), ()); }
    fn is_captured(&self, npc_id: ProtoNpcId, idx: u32) -> bool { self.captured.contains_key(&(npc_id, idx)) }
    fn clear(&self) { self.captured.clear(); }
}
```

### bevy_tasker — offload CPU work to thread/worker pool

`bevy_tasker` is already used in the isometric game for procedural terrain chunk generation (`tilemap.rs`). Same pattern applies to creature systems.

**Where bevy_tasker fits:**

| Work Item | Current | With bevy_tasker |
|-----------|---------|-----------------|
| Broadcast serialization | Clone message N times on main thread | `spawn()` serialization task, fan out pre-serialized bytes |
| Creature slot assignment | `assign_firefly_slots` scans chunks per-frame on main thread | `spawn()` chunk scan, send assignments via channel (like tilemap mesh gen) |
| NpcDb loading | Synchronous `build_creature_registry()` at startup | `spawn()` async load from proto binary or JSON, insert resource on completion |
| Capture validation | Inline in `process_creature_captures()` | `spawn()` validation + registry lookup, main thread just applies result |
| Batch captured state | Iterate HashSet + send N messages per new client | `spawn()` pre-build batch payload, main thread sends single message |

**Pattern (already in codebase):**
```rust
// From tilemap.rs — same pattern works for creature work
let (tx, rx) = crossbeam_channel::unbounded();
bevy_tasker::spawn(async move {
    let result = expensive_computation();
    let _ = tx.send(result);
}).detach();

// In a Bevy system, drain the channel:
while let Ok(result) = rx.try_recv() {
    // apply result to ECS
}
```

**Platform support:**
- Desktop: crossbeam thread pool (1 worker per CPU core)
- WASM+atomics: SharedArrayBuffer web worker pool
- WASM without atomics: falls back to browser microtask queue (single-threaded)

### Combined pattern: DashMap + bevy_tasker

For maximum throughput, combine both:
1. `bevy_tasker::spawn()` offloads capture validation to a worker
2. Worker reads `DashMap<(ProtoNpcId, u32), ()>` concurrently to check duplicates
3. Worker inserts into DashMap on success
4. Worker sends result (captor_id, broadcast message) back via channel
5. Main thread drains channel, sends broadcast to lightyear senders

This keeps the main ECS tick free from capture validation latency while DashMap provides lock-free concurrent access to the shared capture state.

---

## Summary

| Issue | Severity | Type | Effort |
|-------|----------|------|--------|
| Box::leak() in NpcDb::insert() | **P0** | Memory leak | Low |
| Broadcast clone amplification | **P0** | CPU (scales with players) | Medium |
| DefaultHasher instability in ProtoNpcId | **P1** | Correctness risk | Low |
| CapturedCreatures per-client iteration | **P1** | Algorithmic O(N×M) | Medium |
| String cloning in NpcDb::insert() | **P2** | Wasteful allocation | Low |
| find_by_* returning Vec | **P2** | API design | Low |
| .into() allocations in registry builder | **P3** | Startup cost only | Low |
| Linear creature pool scan | **P3** | Acceptable at current scale | Low |
| DashMap for concurrent creature state | **P2** | Concurrency enabler | Medium |
| bevy_tasker for broadcast/assignment offload | **P1** | Main-thread latency | Medium |

## File Reference

| Area | Key Files |
|------|-----------|
| NPC registry | `packages/rust/bevy/bevy_npc/src/registry.rs` |
| Creature ECS types | `packages/rust/bevy/bevy_npc/src/creature.rs` |
| Creature registry + slot helpers | `packages/rust/bevy/bevy_kbve_net/src/npcdb.rs` |
| Wire protocol | `packages/rust/bevy/bevy_kbve_net/src/protocol.rs` |
| Client networking | `apps/kbve/isometric/src-tauri/src/game/net.rs` |
| Server gameserver | `apps/kbve/axum-kbve/src/gameserver/mod.rs` |
| bevy_tasker | `packages/rust/bevy/bevy_tasker/src/lib.rs` |
| Existing tasker usage | `apps/kbve/isometric/src-tauri/src/game/tilemap.rs` |

**Comment by h0lybyte** [comment]
Resolved by #10559 (merged to dev 2026-05-07). All P0/P1/P2 items + P3 item 1 (O(1) entity-bound check via HashSet mirror in CreatureIdIndex) closed. Slot helpers (P3 item 2) untouched per the issue's own "already optimal" verdict.
